tiatoolbox/tools/graph.py

```diff
@@ -0,0 +1,275 @@
+    for c in unique_clusters:
+        (idx,) = np.where(clusters == c)
+        # Find the xy and feature space averages of the cluster
+        point_centroids.append(np.round(points[idx, :].mean(axis=0)))
+        feature_centroids.append(features[idx, :].mean(axis=0))
+    point_centroids = np.array(point_centroids)
+    feature_centroids = np.array(feature_centroids)
+
+    adjacency_matrix = delaunay_adjacency(
+        points=point_centroids,
+        dthresh=connectivity_distance,
+    )
+    edge_index = affinity_to_edge_index(adjacency_matrix)
+
+    result = {
+        "x": feature_centroids,
+        "edge_index": edge_index,
+        "coords": point_centroids,
+    }
+    if label is not None:
+        result["y"] = np.array([label])
+    return result
+
+
+def delaunay_adjacency(points: ArrayLike, dthresh: Number) -> ArrayLike:
+    """Create an adjacency matrix via Delaunay triangulation from a list of coordinates.
+
+    See https://en.wikipedia.org/wiki/Adjacency_matrix.
+
+    Args:
+        coordinates (ArrayLike): A nxm list of coordinates.
+        dthresh (int): Distance threshold for triangulation.
+
+    Returns:
+        ArrayLike: Adjacency matrix of shape NxN where 1 indicates
+            connected and 0 indicates unconnected.
+
+    Example:
+        >>> points = np.random.rand(100, 2)
+        >>> adjacency = delaunay_adjacency(points)
+
+    """
+    # Validate inputs
+    if not isinstance(dthresh, Number):
+        raise TypeError("dthresh must be a number.")
+    if len(points) < 4:
+        raise ValueError("Points must have length >= 4.")
+    if len(np.shape(points)) != 2:
+        raise ValueError("Points must have an NxM shape.")
+    # Apply Delaunay triangulation to the coordinates to get a
+    # tessellation of triangles.
+    tessellation = Delaunay(points)
+    # Find all connected neighbours for each point in the set of
+    # triangles. Starting with an empty dictionary.
+    traingle_neighbours = defaultdict(set)
+    # Iterate over each triplet of point indexes which denotes a
+    # triangle within the tessellation.
+    for index_triplet in tessellation.simplices:
+        for index in index_triplet:
+            connected = set(index_triplet)
+            connected.remove(index)  # Do not allow connection to itself.
+            traingle_neighbours[index] = traingle_neighbours[index].union(connected)
+    # Initialise the nxn adjacency matrix with zeros.
+    adjacency = np.zeros((len(points), len(points)))
+    # Fill the adjacency matrix:
+    for index in traingle_neighbours:
+        neighbours = traingle_neighbours[index]
+        neighbours = np.array(list(neighbours), dtype=int)
+        kdtree = cKDTree(points[neighbours, :])
+        nearby_neighbours = kdtree.query_ball_point(
+            x=points[index],
+            r=dthresh,
+        )
+        neighbours = neighbours[nearby_neighbours]
+        adjacency[index, neighbours] = 1.0
+        adjacency[neighbours, index] = 1.0
```

Comment:
This part doesn't seem to be Delaunay to me. What is kdtree doing here? @wenqi006

Reply:
The KDTree is used to only connect neighbours within a certain distance. This can be done without a KDTree but for many points it if faster to use a KDTree. You make a good point, though. This step could possibly be skipped if the distance threshold is set to 0 or a negative number to mean no threshold on maximum distance to connect.

Reply:
@vqdang @wenqi It just occurred to me that the number of neighbours to check here is quite small. We could replace the kdtree with a simple list enumeration e.g.:

```python
nearby_neighbours = [n for n in neighbours if np.linalg.norm(n, points[index]) < dthresh]
```